### PR TITLE
Assert value check's replaced by error's

### PR DESCRIPTION
**Pull request by jdziurzynski**
Issue : KNTK-40

Replaced "assert" statements with suitable exceptions e.g. `IncompleteObjectError`. "assert" statements evaluates only in debug mode.

### CODE REVIEW

File: kentik_api_library/kentik_api/api_resources/custom_dimensions_api.py [hunk 1]
```diff
@@ -97,9 +103,12 @@ def get_all(self) -> List[CustomDimension]:
         return custom_dimensions_payload.GetAllResponse.from_json(response.text).to_custom_dimensions()
 
     def create(self, custom_dimension: CustomDimension) -> CustomDimension:
-        assert custom_dimension.name is not None
-        assert custom_dimension.display_name is not None
-        assert custom_dimension.type is not None
+        if custom_dimension.name is None:
+            raise IncompleteObjectError("Custom Dimensions", "custom dimension name value is required")
```
Comment (danielfurman) [inline kentik_api_library/kentik_api/api_resources/custom_dimensions_api.py:107]:
```suggestion
            raise IncompleteObjectError("Custom Dimensions", "name is required")
```
In other places too.

**Review by danielfurman — CHANGES_REQUESTED** [review]

File: kentik_api_library/kentik_api/api_resources/custom_applications_api.py [hunk 2]
```diff
@@ -17,7 +18,8 @@ def get_all(self) -> List[CustomApplication]:
         return custom_applications_payload.GetAllResponse.from_json(response.text).to_custom_applications()
 
     def create(self, custom_application: CustomApplication) -> CustomApplication:
-        assert custom_application.name is not None
+        if custom_application.name is None:
+            raise IncompleteObjectError("Custom Aplication", "custom aplication name is required")
```
[the remark this answers is not in the document]
Comment (mmac-m3a) [reply kentik_api_library/kentik_api/api_resources/custom_applications_api.py:22]: Consider refactoring the exception to take class instance as the first argument. With that you can use `obj.__class__.__name__` to get the class name. So, the code would look like this:

```
raise IncompleteObjectError(custom_application, "name is required")
```

We may also consider meta-programming the validation by giving each class a list of constraints. Something like this (using `CustomApplication` as an example):

```python
class CustomApplication:
    CONSTRAINTS = [
        {"attr": "name", "validator": lambda x: x is not None, "msg": "name is required"} 
    ]
   ...

def validate_object(obj: Any):
    if not hasattr(obj, CONSTRAINTS):
        return
    for c in obj.CONSTRAINTS:
       if not hasattr(obj, c["attr"]):
           raise InvalidConstraint(c, f"class {c.__class__.__name__} has no attribute {c['attr']")
       if not c["validator"](getattr(obj, c["name"):
           raise IncompleteObjectError(obj, c["msg"])
``` 
(we will obviously need to create the `InvalidConstraint` exception)

The code here will look like this:

```python
def create(self, custom_application: CustomApplication) -> CustomApplication:
    validate_object(custom_application)
    ...
```

[the remark this answers is not in the document]
Comment (danielfurman) [reply kentik_api_library/kentik_api/api_resources/custom_applications_api.py:22]: @mmac-m3a I created a task for making validation code consistent across resources (https://clkentik.atlassian.net/browse/KNTK-386). Your suggestion could be applied as a part of that. However for me it seems too magic.

[the remark this answers is not in the document]
Comment (jdziurzynski) [reply kentik_api_library/kentik_api/api_resources/custom_applications_api.py]: After conversation with Daniel and Mateusz, I created a task regarding to exceptions https://clkentik.atlassian.net/browse/KNTK-388

File: kentik_api_library/kentik_api/api_resources/custom_dimensions_api.py [hunk 3]
```diff
@@ -78,6 +75,15 @@ def delete(self, custom_dimension_id: ID, populator_id: ID) -> bool:
         response = self.send(apicall)
         return response.http_status_code == HTTPStatus.NO_CONTENT
 
+    @staticmethod
+    def check_fields(populator: Populator):
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply kentik_api_library/kentik_api/api_resources/custom_dimensions_api.py]: Ticket: https://clkentik.atlassian.net/browse/KNTK-386

File: kentik_api_library/kentik_api/api_resources/custom_dimensions_api.py [hunk 4]
```diff
@@ -78,6 +75,15 @@ def delete(self, custom_dimension_id: ID, populator_id: ID) -> bool:
...
+        if populator.value is None:
+            raise IncompleteObjectError("Populators", "value is required")
```
[the remark this answers is not in the document]
Comment (danielfurman) [reply kentik_api_library/kentik_api/api_resources/custom_dimensions_api.py]: Done

File: kentik_api_library/kentik_api/api_resources/custom_dimensions_api.py (hunk 1 above)
Comment (danielfurman) [reply kentik_api_library/kentik_api/api_resources/custom_dimensions_api.py]: Done

**Review by danielfurman — APPROVED** [review]

File: kentik_api_library/kentik_api/api_resources/custom_applications_api.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mmac-m3a) [reply kentik_api_library/kentik_api/api_resources/custom_applications_api.py]: I don't quite see what is magic about having validation constraints associated with data representation. There are surely different ways to implement it and I by no means insist on the very code I've proposed. I though prefer a solution that eliminates repetitive code and ideally makes validations logic easy to understand.

**Review by mmac-m3a — DISMISSED** [review]

**Comment by mmac-m3a** [comment]
I'm glad to see that our CI tests are more careful than me :). "String" or "string" are really not valid Python types (or objects in the app). Please, run `mypy` (`python3 setup.py mypy`) on the repo and fix all problems.

**Review by danielfurman — APPROVED** [review]

**Review by mmac-m3a — APPROVED** [review]

**Pull request merged**